kernel/src/hil/usb_hid.rs

```diff
@@ -0,0 +1,86 @@
+        len: usize,
+        endpoint: usize,
+    );
+
+    /// Called when a packet has been finished transmitting.
+    /// This will return the buffer passed into `send_buffer()` as well as
+    /// the endpoint where the data was sent. The `len` argument contains
+    /// the number of bytes sent. If this is less then the size of the
+    /// buffer `result` will contain the `ESIZE` error.
+    fn packet_transmitted(
+        &'a self,
+        result: ReturnCode,
+        buffer: &'static mut [u8; 64],
+        len: usize,
+        endpoint: usize,
+    );
+
+    /// Called when checking if we can start a new receive operation.
+    /// Should return true if we are ready to receive and not currently
+    /// in the process of receiving anything. That is if we are currently
+    /// idle.
+    /// If there is an outstanding call to receive, a callback already
+    /// waiting to be called then this will return false.
+    fn can_receive(&'a self) -> bool;
+}
+
+pub trait UsbHid<'a> {
+    /// Sets the buffer to be sent and starts a send transaction.
+    /// On success returns the number of bytes that will be sent.
+    /// On failure returns an error code and the buffer passed in.
+    fn send_buffer(
+        &'a self,
+        send: &'static mut [u8; 64],
+    ) -> Result<usize, (ReturnCode, &'static mut [u8; 64])>;
+
+    /// Cancels a send called by `send_buffer()`. Once called the
+    /// `packet_transmitted()` callback won't occur and no new data
+    /// will be sent. Any data currently being sent will stil be sent.
+    /// On success (a send was cancelled) returns the buffer passed to
+    /// `send_buffer()`.
+    /// On failure returns a return code
+    fn send_cancel(&'a self) -> Result<&'static mut [u8; 64], ReturnCode>;
+
+    /// Sets the buffer for received data to be stored and enables receive
+    /// transactions. Once this is called the implementation will enable
+    /// receiving via USB. Once a packet is received the `packet_received()`
+    /// callback will be triggered and no more data will be received until
+    /// this is called again.
+    ///
+    /// Once this is called, the implementation will wait until either
+    /// a packet is received or `receive_cancel()` is called.
+    ///
+    /// Calling `receive_buffer()` while there is an outstanding
+    /// `receive_buffer()` operation will return EBUSY.
+    ///
+    /// On success returns nothing.
+    /// On failure returns an error code and the buffer passed in.
+    fn receive_buffer(
+        &'a self,
+        recv: &'static mut [u8; 64],
+    ) -> Result<(), (ReturnCode, &'static mut [u8; 64])>;
+
+    /// Cancels a receive called by `receive_buffer()`. Once called the
+    /// `packet_received()` callback won't occur and no data will be
+    /// received until `receive_buffer()` is called.
+    /// On success (a receive was cancelled) returns the buffer passed to
+    /// `receive_buffer()`.
+    /// On failure returns a return code
```

Comment:
I think the cancellation semantics are cleaner here, only question is the "on failure returns a return code" --> will all failures still results in eventual callbacks later? What return codes are valid here?

Reply:
The return codes should be specified as well as that a callback will be invoked.

Note one issue here: what if there is no call outstanding? You don't want to return a buffer and you don't want to signal a callback. This is why the radio moved towards cancel basically always invoking a callback: SUCCESS meant there was no outstanding call. This also simplified buffer management as there's a single call path to get the buffer back.

Reply:
I have updated this to match the `send_cancel()`